Support onion message pathfinding

valentinewallace opened the pull request:
Take 2, Github wouldn't let me reopen #1669 🤔 

We could reuse router::get_route, but it's better to start from scratch
because get_route includes a lot of payment-specific computations that impact
performance.

- [x] check for OM feature bit
- [x] finish tests

Partially addresses https://github.com/lightningdevkit/rust-lightning/issues/1607

~Based on #1731~ 

Known followups:
* scoring
* benchmarking (will need to update the netgraph file to support the OM feature bit)

valentinewallace commented:
I think I addressed all the feedback from #1669. Much cleaner! 

If there are rough conceptual ACKs, I can add more testing, add feature bit checks, and open the first commit in a separate PR.

TheBlueMatt reviewed (commented):
At first glance it looks like it'll work. We definitely need some ability to score nodes, if only a "skip nodes" set, as we'll want the ability to, say, fetch two non-overlapping paths.

valentinewallace commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,176 @@
+						// We may push a given successor multiple times, but the heap should sort its best entry
+						// to the top. We do this because there is no way to adjust the priority of an existing
+						// entry in `BinaryHeap`.
+						frontier.push(PathBuildingHop {
```
Yeah, it seems slower. 

Not checking if `successor` is in `visited`:
```
test routing::onion_message::benches::generate_routes_test                   ... bench:  27,532,191 ns/iter (+/- 13,250,916)
test routing::onion_message::benches::generate_routes_test                   ... bench:  25,474,037 ns/iter (+/- 13,319,206)
test routing::onion_message::benches::generate_routes_test                   ... bench:  27,410,199 ns/iter (+/- 15,179,713)
```

Checking:
```
test routing::onion_message::benches::generate_routes_test                   ... bench:  32,435,508 ns/iter (+/- 11,682,837)
test routing::onion_message::benches::generate_routes_test                   ... bench:  30,536,545 ns/iter (+/- 11,874,035)
test routing::onion_message::benches::generate_routes_test                   ... bench:  31,512,712 ns/iter (+/- 12,148,234)
```

valentinewallace committed:
Benchmark OM pathfinding

valentinewallace commented:
> We definitely need some ability to score nodes, if only a "skip nodes" set, as we'll want the ability to, say, fetch two non-overlapping paths.

Sounds good, will likely do it in follow-up if that works

valentinewallace committed:
OM pathfinding: don't consider nodes that don't support OM forwarding

Or that require unknown features

valentinewallace committed:
Move get_channel_details helper into router::test_utils

Used incoming commit(s) to test onion message pathfinding

valentinewallace committed:
Implement pathfinding for onion messages

We *could* reuse router::get_route, but it's better to start from scratch
because get_route includes a lot of payment-specific computations that impact
performance.

naumenkogs commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,185 @@
+impl PartialOrd for PathBuildingHop {
+	fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+		// We need a min-heap, whereas `BinaryHeap`s are a max-heap, so compare the costs in reverse.
+		other.cost.partial_cmp(&self.cost)
```
Assuming the cost is equal, can someone grind lower node-id to attract more message flows? After staring at rust documentation for 5 minutes, I'm not sure.

valentinewallace replied:
Added a TODO for this

naumenkogs replied:
So you confirm it's possible by looking at the rust doc?

naumenkogs replied:
Perhaps fine to keep a todo anyway because:
- the way nodes sorted before insertion into heap might play a role then
- heap implementation may change

naumenkogs commented on lightning/src/routing/onion_message.rs:
```diff
@@ -60,6 +60,9 @@ pub fn find_path<L: Deref, GL: Deref>(
 		if let Some(node_info) = network_nodes.get(&node_id) {
+			if let Some(node_ann) = &node_info.announcement_info {
+				if !node_ann.features.supports_onion_messages() { continue; }
+			} else if node_id != our_node_id { continue; }
```
I think this line is particularly hard to understand, could you add a comment? Or even, do this check 3 lines before even accessing `announcement_info`.

valentinewallace replied:
Lmk if the new version + comment make sense

valentinewallace committed:
OM pathfinding: don't consider offline network successors

As judged by whether the channel is disabled, similar to payment routing

valentinewallace committed:
OM pathfinding: ensure first_hops overrides network graph

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,328 @@
+	}
+
+	path.reverse();
+	log_info!(logger, "Got route to {:?}: {:?}", destination, path);
```
As `reverse_path` is kind of a helper function, maybe this log could be placed in `find_route` itself?

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,328 @@
+		if let Some(node_info) = network_nodes.get(&node_id) {
+			// Only consider the network graph if first_hops does not override it.
+			if valid_first_hops.contains(&node_id) || node_id == our_node_id {
+			} else if let Some(node_ann) = &node_info.announcement_info {
```
IIUC this is only to account for the case where the first hop features are different than the network-announced ones? What would be a case in which we can expect that?

Also, this currently seems to default to `supports_onion_messages == true` in case we haven't gotten a node announcement. Maybe this should rather be pessimistic and exclude a hop that may not support onion messages?

valentinewallace replied:
> Also, this currently seems to default to supports_onion_messages == true in case we haven't gotten a node announcement. Maybe this should rather be pessimistic and exclude a hop that may not support onion messages?

Indeed, that was a regression 🙈 

> IIUC this is only to account for the case where the first hop features are different than the network-announced ones? What would be a case in which we can expect that?

Basically our local channel view is expected to be the most up-to-date. Gossip can be slow but we'll be the first to know about feature upgrade/downgrades for our channel peers

tnull replied:
> Basically our local channel view is expected to be the most up-to-date. Gossip can be slow but we'll be the first to know about feature upgrade/downgrades for our channel peers

Ah, I see. Makes sense!

tnull reviewed (commented):
First pass, looks good! Haven't looked at tests yet.

valentinewallace commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,328 @@
+	our_node_pubkey: &PublicKey, destination: &PublicKey, network_graph: &NetworkGraph<GL>, first_hops: Option<&[&ChannelDetails]>, logger: L
+) -> Result<Vec<PublicKey>, Error> where L::Target: Logger, GL::Target: Logger
+{
+	log_trace!(logger, "Searching for an onion message path from origin {} to destination {} and {} first hops {}overriding the network graph",
```
Note that the current version matches the equivalent log in `get_route`. We could do that but IIUC we'd have to add a call to `format!` to insert the number of first_hops, which allocates a String, which doesn't seem worth?

tnull replied:
Fair enough, probably not worth.

valentinewallace commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,328 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Onion message pathfinding lives here.
```
Lmk if the new docs look good

tnull replied:
Generally looks good. Is the current mix of `//!`, `///`, and `//` as intended though?

valentinewallace commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,328 @@
+
+#[derive(Debug, PartialEq)]
+/// Errored running [`find_path`].
+pub enum Error {
```
I didn't do `Error` because I'm not sure what implementing `source` would look like here, but did `Display`

tnull replied:
I believe when the error already implements `Display` and `Debug`, you should be fine with

```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {}
```

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,283 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! [Onion message] pathfinding lives here. Finding paths for onion messages is necessary for
```
nit: If you put the first sentence in its own paragraph, it will be the only one shown in  the 'Modules' overview of the [`lightning::routing` docs](https://docs.rs/lightning/latest/lightning/routing/index.html).
```suggestion
//! [Onion message] pathfinding lives here.
//!
//! Finding paths for onion messages is necessary for
```

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,283 @@
+//! [Onion message]: crate::onion_message
+//! [offers]: <https://github.com/lightning/bolts/pull/798>
+
+// Note that we do not reuse `router::find_route` because it contains a lot of payment-specific
```
nit: Not sure we need this comment anymore since you already explained it in more detail above?

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,283 @@
+	}
+
+	#[test]
+	fn three_hops() {
```
Not sure whether this test case is not a superset of `one_hop`/`two_hops`? Do we need all three, or can we just keep this one?

tnull commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,283 @@
+		assert_eq!(path.len(), 1);
+		assert_eq!(path[0], node_pks[0]);
+	}
+}
```
Maybe test cases for a) checking that valid first hops are included and b) that invalid first hops are not considered (cf. `router::tests::invalid_first_hop_test`) might be nice.

tnull commented on lightning/src/routing/test_utils.rs:
```diff
@@ -131,7 +167,7 @@ pub(super) fn id_to_feature_flags(id: u8) -> Vec<u8> {
 }
 
-pub(super) fn build_line_graph() -> (
+pub(super) fn build_line_graph(node_features: NodeFeatures) -> (
```
nit: Rename to make it consistent:
```suggestion
pub(super) fn build_line_graph_with_features(node_features: NodeFeatures) -> (
```

tnull reviewed (commented):
Mostly focused on tests this time.

valentinewallace committed:
OM pathfinding: add our id and first hop checks

Don't allow finding a path to ourselves and reject first hops that contain our
node id as the counterparty node id.

TheBlueMatt reviewed (commented):
Basically LGTM, one real comment.

ariard commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,280 @@
+//! Finding paths for onion messages is necessary for retrieving invoices and fulfilling invoice
+//! requests from [offers]. It differs from payment pathfinding in that channel liquidity, fees, and
+//! knobs such as `htlc_maximum_msat` do not factor into path selection -- onion messages require a
+//! peer connection and nothing more. However, we still use the network graph because onion messages
+//! between channel peers are likely to be prioritized over those between non-channel peers.
```
So while I think relying on the existence of channels between peers to flow onion messages align well the incentives between bandwidth consumed and onion forwards accepted, I don't know if this is a direction we would like for the long-term in term of privacy. Indeed, overlapping both offers flows and HTLC flows, if we re-use the same heuristics than for classic pathfinding we might have overlap and show more "data trails" to any payment observers. Especially onion messages routing might invalidate some privacy-preserving enhancing we have implemented with HTLCs such as https://github.com/lightningdevkit/rust-lightning/pull/1286. The lack of timing padding in onion routing might cancel the CLTV delta offset. Of course, this is a subject we can think more and improve with time. Just to be aware of about the privacy regression we might have.

ariard reviewed (commented):
Nice if we can this back, figured out gonna be needed for #1848 and beyond. Will review soon.

valentinewallace commented:
> Nice if we can this back, figured out gonna be needed for #1848 and beyond. Will review soon.

Thanks, I don't know if I'll be prioritizing the follow-ups for a while (open to someone taking them over), but since this seems close(?) I'll try to get back to it in the new year

ariard commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,280 @@
+	let start = NodeId::from_pubkey(&our_node_pubkey);
+	let mut valid_first_hops = HashSet::new();
+	let mut frontier = BinaryHeap::new();
+	frontier.push(PathBuildingHop { cost: 0, node_id: start, parent_node_id: start });
```
Can define the reasoning behind `PathBuildingHop::cost`. I don't know if the onion bandwidth we're expecting from hops on the network is scaled on the channel capacity. The most rule of thumb metric I would guess.

ariard reviewed (commented):
Beyond offer transport, I think onion messages pathfinding could be useful to two more use-cases: bitcoin transaction relay and jamming credentials. For the former, I believe the set of heuristics is going to be really-specific, as you might avoid centralised hops to improve fault-tolerance. For the latter, the use-case requirements are more likely to be similar to offers, however you might do synthetic traffic to mask the economic volume.

Just to say, I don't think we'll need more than one onion messages pathfinding algorithms for other use-cases, however I believe they might need each to plug a `CustomHeuristicsOnionScore` similar to `scoring::Score`, or combine those `CustomHeuristicsOnionScore`. Thinking aloud for jamming.

ariard commented:
@naumenkogs if you're bored, I'm pretty sure onion message pathfinding will be a must to play with a bunch of things, notably jamming mitigations :)

valentinewallace committed:
WIP: try pushing scids on the heap instead

valentinewallace commented on lightning/src/routing/onion_message.rs:
```diff
@@ -0,0 +1,280 @@
+				{ continue; }
+			} else { continue; }
+			for scid in &node_info.channels {
+				if let Some(chan_info) = network_channels.get(&scid) {
```
I'm getting somewhat comparable times, with scid-heap narrowly taking the lead.

inserting node_ids into heap:
```
test routing::onion_message::benches::generate_simple_routes                 ... bench:  42,412,441 ns/iter (+/- 26,718,082)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  40,244,529 ns/iter (+/- 27,297,264)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  38,865,025 ns/iter (+/- 27,147,061)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  43,582,987 ns/iter (+/- 23,421,789)
```

inserting scids into heap:
```
test routing::onion_message::benches::generate_simple_routes                 ... bench:  38,450,779 ns/iter (+/- 29,381,214)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  37,930,562 ns/iter (+/- 27,023,358)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  41,410,850 ns/iter (+/- 27,262,014)
test routing::onion_message::benches::generate_simple_routes                 ... bench:  39,008,820 ns/iter (+/- 26,009,606)
```

I pushed a WIP commit (https://github.com/lightningdevkit/rust-lightning/pull/1724/commits/28318ac76ca598ff9f37709202b9904dd7e9838a) with the scid-heap version if you want to test it out and/or check that I implemented it the way you expected.

valentinewallace commented:
I see @moneyball added this to the 119 milestone.

But I don't see the rush on this, because most cases will use direct-connect? It seems like this will only be useful once OMs are more widely supported across the network. @TheBlueMatt @jkczyz Let me know if I'm good to remove this from the milestone for now.

moneyball commented:
Yes I'll remove it. I had added it to 119 due to a misunderstanding of how it could work in the short-term.
